labs/RivePlayer/src/RivePlayer.cs

```diff
@@ -0,0 +1,255 @@
+    private List<Action>? _deferredSMInputsDuringAsyncSourceLoad = null;
+
+    private async void LoadSourceFileDataAsync(string uriString, int sourceToken)
+    {
+        if (Uri.TryCreate(uriString, UriKind.Absolute, out var uri))
+        {
+            Stream? stream = null;
+            if (uri.Scheme == "http" || uri.Scheme == "https")
+            {
+                try
+                {
+                    HttpResponseMessage response = await new HttpClient().GetAsync(uri);
+                    if (response.IsSuccessStatusCode)
+                    {
+                        stream = await response.Content.ReadAsStreamAsync();
+                    }
+                }
+                catch (HttpRequestException e)
+                {
+                    // TODO: Load a 404 file?
+                    Console.WriteLine("Failed to connect to " + uri.ToString());
+                    Console.WriteLine(e.ToString());
+                }
+            }
+            else if (uri.Scheme == "ms-appx")
+            {
+                var file = await StorageFile.GetFileFromApplicationUriAsync(uri);
+                if (file != null && sourceToken == _currentSourceToken)
+                {
+                    IBuffer buffer = await FileIO.ReadBufferAsync(file);
+                    stream = buffer.AsStream();
+                }
+            }
+            if (stream != null && sourceToken == _currentSourceToken)
+            {
+                byte[] data = new byte[stream.Length];
+                stream.Read(data, 0, data.Length);
+                stream.Dispose();  // Don't keep the file open.
+                sceneActionsQueue.Enqueue(() => UpdateScene(SceneUpdates.File, data));
+                // Apply deferred state machine inputs once the scene is fully loaded.
+                foreach (Action stateMachineInput in _deferredSMInputsDuringAsyncSourceLoad!)
+                {
+                    sceneActionsQueue.Enqueue(stateMachineInput);
+                }
+            }
+        }
+        _deferredSMInputsDuringAsyncSourceLoad = null;
+    }
+
+    private void EnqueueStateMachineInput(Action stateMachineInput)
+    {
+        if (_deferredSMInputsDuringAsyncSourceLoad != null)
+        {
+            // A source file is currently loading async. Don't set this input until it completes.
+            _deferredSMInputsDuringAsyncSourceLoad.Add(stateMachineInput);
+        }
+        else
+        {
+            sceneActionsQueue.Enqueue(stateMachineInput);
+        }
+    }
+
+    public void SetBool(string name, bool value)
+    {
+        EnqueueStateMachineInput(() => _scene.SetBool(name, value));
+    }
+
+    public void SetNumber(string name, float value)
+    {
+        EnqueueStateMachineInput(() => _scene.SetNumber(name, value));
+    }
+
+    public void FireTrigger(string name)
+    {
+        EnqueueStateMachineInput(() => _scene.FireTrigger(name));
+    }
+
+    private delegate void PointerHandler(Vec2D pos);
```

Comment:
Is a custom delegate really needed here? Wouldn't `Action<Vec2D>` suffice?

Reply:
TIL again. That's cool!